ddtrace/encoding.py

```diff
@@ -44,42 +45,78 @@ def encode_traces(self, traces):
         :param traces: A list of traces that should be serialized
         """
         normalized_traces = [[span.to_dict() for span in trace] for trace in traces]
-        return self._encode(normalized_traces)
+        return self.encode(normalized_traces)
 
-    def encode_services(self, services):
+    def encode_trace(self, trace):
         """
-        Encodes a dictionary of services.
+        Encodes a traces, expecting a list of spans. Before dump the string in a
+        serialized format all traces are normalized, calling the ``to_dict()`` method.
+        The traces nesting is not changed.
 
-        :param services: A dictionary that contains one or more services
+        :param trace: A list of traces that should be serialized
         """
-        return self._encode(services)
+        return self.encode([span.to_dict() for span in trace])
 
-    def _encode(self, obj):
+    def encode(self, obj):
         """
         Defines the underlying format used during traces or services encoding.
         This method must be implemented and should only be used by the internal functions.
         """
         raise NotImplementedError
 
+    def decode(self, data):
+        """
+        Defines the underlying format used during traces or services encoding.
+        This method must be implemented and should only be used by the internal functions.
+        """
+        raise NotImplementedError
+
+    def finish(self, objs):
+        """
+        """
+        raise NotImplementedError
+
 
 class JSONEncoder(Encoder):
     def __init__(self):
         # TODO[manu]: add instructions about how users can switch to Msgpack
         log.debug('using JSON encoder; application performance may be degraded')
         self.content_type = 'application/json'
 
-    def _encode(self, obj):
+    def encode(self, obj):
         return json.dumps(obj)
 
+    def decode(self, data):
+        return json.loads(data)
+
+    def finish(self, objs):
+        return '[' + ','.join(objs) + ']'
+
 
 class MsgpackEncoder(Encoder):
     def __init__(self):
         log.debug('using Msgpack encoder')
         self.content_type = 'application/msgpack'
 
-    def _encode(self, obj):
+    def encode(self, obj):
         return msgpack.packb(obj, **MSGPACK_PARAMS)
 
+    def decode(self, data):
+        return msgpack.unpackb(data)
+
+    def finish(self, objs):
+        buf = b''.join(objs)
+
+        # Prepand array header to buffer
+        # https://github.com/msgpack/msgpack-python/blob/f46523b1af7ff2d408da8500ea36a4f9f2abe915/msgpack/fallback.py#L948-L955
```

Comment:
I actually think keeping this here is better, at least until we vendor `msgpack`, since not all versions of `msgpack` have this helper function.

But also, this is exactly the code that `pack_array_header` calls, and the same code exists in `dd-trace-js` and `dd-trace-go`.